src/rust/engine/src/core.rs

```diff
@@ -351,23 +368,88 @@ impl Failure {
       }
     }
   }
-}
 
-impl Failure {
+  ///
+  /// When an exception is raised from an inner `await Get(...)` and handled in a parent coroutine,
+  /// but then an exception is raised after handling it, the traceback of the second exception needs
+  /// to be joined with the first to present the complete traceback history of the raised exception
+  /// to the user.
+  ///
+  pub fn join_tracebacks(self, new_failure: Self) -> Failure {
+    match (self, new_failure) {
+      (
+        Failure::Throw {
+          val: _old_val,
+          python_traceback: old_python_traceback,
+          engine_traceback: old_engine_traceback,
+        },
+        Failure::Throw {
+          val: new_val,
+          python_traceback: new_python_traceback,
+          engine_traceback: new_engine_traceback,
+        },
+      ) => {
+        let joined_python_tracebacks = [
+          old_python_traceback,
+          "While handling that, another exception was raised:\n".to_string(),
+          new_python_traceback,
+        ]
+        .join("\n");
+        let joined_engine_tracebacks = old_engine_traceback
+          .into_iter()
+          .chain(new_engine_traceback.into_iter())
+          .collect::<Vec<_>>();
+        Failure::Throw {
+          val: new_val,
+          python_traceback: joined_python_tracebacks,
+          engine_traceback: joined_engine_tracebacks,
+        }
+      }
+      (_, new) => new,
+    }
+  }
+
+  ///
+  /// Get a handle to the exception Value, if applicable, or wrap this in an InternalFailure.
+  ///
+  pub fn as_py_err(&self) -> Result<&Value, InternalFailure> {
```

Comment:
While `Failure::Invalidated` does not have an associated python exception, it would still be possible to generate one here (by calling roughly `throw(&e.to_string())` or something related for `Invalidated`). So I don't think that the "InternalFailure" wrapper is strictly necessary.

It should not be the case that an invalidation made it to the root of a run, but if it did, you would still want to render it as a python exception. It looks like that's all the code that handles `InternalFailure` does with it in any case.

Reply:
I would love to justify removing this code so this is a great argument to hear. I will try that.

Reply:
Done!